### PR TITLE
[AppRewrite] Part 4: Telestion Loader

### PR DESCRIPTION
**Pull request by fussel178**
### Stacked Pull Request

This is a stacked pull request. Please wait on pending pull requests that reference on this branch before merging this branch.
Take a look at https://www.michaelagreiler.com/stacked-pull-requests/ for more information.

### Summary 

This PR introduces a reference implementation for a loader.

#### The Telestion Loader

The Telestion loader provides a reference implementation of a loader so you don't need to implement all event hooks over and over again when you want to develop your own loader.

Let's say, you want to deploy some verticles after the Vert.x startup:

```java
public class MyLoader extends TelestionLoader<GenericLoaderConfiguration> {

	@Override
	public void onAfterVertxStartup(Promise<Void> startPromise) throws Exception {
		CompositeFuture.join(
			launcher.deploy(new PositionPublisher()),
			launcher.deploy(DebugVerticle.class, new DeploymentOptions()),
			launcher.deploy("some.where.far.away.Verticle")
		).onSuccess(result -> startPromise.succeed()).onFailure(startPromise::fail);
	}
}
```

This loader deploys 3 verticles once the Vert.x instance has finished its startup procedure.

The Telestion Loader provides some convenience methods and allows access to the associated launcher instance. (e.g. for deployment or running other loaders, etc.)

Loaders also accept a configuration record following the same principle as the verticle configuration.

### Sample project

You can find a sample project that uses the new application structure here:

https://github.com/wuespace/telestion-project-playground/tree/fussel178/feat/application-rewrite/application

### Next steps

This PR belongs to a series of changes which updates the application structure in Telestion.

Read next: https://github.com/wuespace/telestion-core/pull/504
Read previous: https://github.com/wuespace/telestion-core/pull/501

### CLA

- [x] I have signed the individual contributor's license agreement **and sent** it to the board of the WüSpace e. V. organization.

### CODE REVIEW

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/GenericLoaderConfiguration.java
Comment (pklaschka) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/GenericLoaderConfiguration.java:14]: Replace with `UntypedLoaderConfiguration` :wink:

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (pklaschka) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:93]: Not a big fan of this warning message, tbh

**Review by pklaschka — CHANGES_REQUESTED** [review]
Apart from a warning message and the fact that we should probably rename `Generic` to `Untyped` here, as well: LGTM!

[cool](https://app.graphite.dev/github/pr/wuespace/telestion-core/502)

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/GenericLoaderConfiguration.java
Comment (cb0s) [reply modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/GenericLoaderConfiguration.java:14]: That 😂

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (cb0s) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:323]: I don't know but I'd rather see the order reversed of public and private reversed. As we are not in C or C++ were we need forwared declaration, I think our silent convention should be to put protected and especially private fields and methods to the bottom, as usually users are only interested in public and maybe protected stuff... I would like to make their life easier and not need them scrolling 100 years 🚀

**Review by cb0s — CHANGES_REQUESTED** [review]
General direction - as usual - right. Some minor things I would like to have changed, most importantly GenericLoader shouldn't be called like that 🤮😂 That feels personal to me :D

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (fussel178) [reply modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:93]: That's interesting, we have this exact same warning message in the `TelestionVerticle` class :thinking: 

https://github.com/wuespace/telestion-core/blob/276285c1ba23d6f86ed22813384bfc8d7dd19ec4/modules/telestion-api/src/main/java/de/wuespace/telestion/api/verticle/TelestionVerticle.java#L62

Comment (fussel178) [reply modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:93]: I think, we should change that, too :see_no_evil:

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (fussel178) [reply modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:323]: Well, that's interesting. We've written the `TelestionVerticle` class in the exact same order :thinking:

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/GenericLoaderConfiguration.java
Comment (fussel178) [reply modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/GenericLoaderConfiguration.java]: Done in [`8cf7509` (#502)](https://github.com/wuespace/telestion-core/pull/502/commits/8cf750925efc880e4438c6f542f64be677842bf5)

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (fussel178) [reply modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java]: Done in [`d2e327a` (#502)](https://github.com/wuespace/telestion-core/pull/502/commits/d2e327a5fbf8031f806fc5f24c7786e58607d26d)

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (pklaschka) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:18]: Language and tone 🙂 

```suggestion
 * The Telestion Loader is a class that implements all required {@link Loader} methods.
 * This provides you a good starting point for your Loader.
 * Overwrite the event handlers you need and do your startup or shutdown steps.
```

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/NoConfiguration.java
Comment (pklaschka) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/NoConfiguration.java:13]: I think this doc could use an example of how to "plug" this configuration type into the loader (i.e., just a simple `Loader` class with this put into the generic field)

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/UntypedConfiguration.java
Comment (pklaschka) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/UntypedConfiguration.java:14]: Same as above: An example on how to apply this could be nice.

**Review by pklaschka — CHANGES_REQUESTED** [review]
Sorry, another few changes 🙂

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (cb0s) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:68]:
```suggestion
	 * Returns <code>null</code> if no type via {@link #getConfigType()} is given.
```
Imo sounds better

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (cb0s) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:79]:
```suggestion
	 * Merges the raw configuration into the default configuration
```

File: modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java
Comment (cb0s) [inline modules/telestion-application/src/main/java/de/wuespace/telestion/application/loader/TelestionLoader.java:136]:
```suggestion
	 * Same as {@link TelestionLoader#TelestionLoader(boolean)}, but enables loading of default configuration
```

**Review by cb0s — CHANGES_REQUESTED** [review]
Some Java docs stuff left before you will get my approval.
So far it looks good though :D

**Pull request closed without merging**
